Fix 9612 and 9825 - Posting: Update primary and secondary options + new Update confirmation dialog

maxme opened the pull request:
This PR supersedes #10041

Fixes #9612 - change the primary / secondary action logic. This is the implementation of the proposition made in paCBwp-95-p2
Fixes #9825 - hide the secondary update button on published posts and create a new confirmation dialog for post updates.

Note: this also fixes the promo dialog that shows on first publish or first schedule. You can test this by resetting the flag with the stetho `dumpapp` script:
```sh
$ ./scripts/dumpapp prefs write org.wordpress.android.beta_preferences ASYNC_PROMO_REQUIRED string true
```

## Test
### Private and Published Posts
- Edit a published or a private post.
- Check there is no secondary push action in the overflow menu.
- Edit the post, tap the Update primary action, notice the dialog.

### Local Drafts and Drafts
- Create or edit a draft.
- Check the primary action is Publish and secondary action is `Save`.
- Edit the post, tap the Publish primary action, notice the dialog.

### Scheduled
- Edit a Scheduled post.
- Check the primary action is `Schedule` and secondary action is `Publish Now`
- Verify the primary and secondary actions.

### Pending
- Edit a Pending post
- Check the primary action is `Save` and secondary action is `Publish Now`
- Verify the primary and secondary actions.

### Trashed
- Edit a Trashed post
- Check the primary action is `Save` and secondary action is `Save As Draft`
- Verify the primary and secondary actions.

## Update release notes:

- I didn't change `RELEASE-NOTES.txt`.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -3986,7 +4025,7 @@ public void onEventMainThread(UploadService.UploadMediaRetryEvent event) {
 
     private boolean shouldShowAsyncPromoDialog() {
```
Note: this function will be re-done or used somewhere else (secondary option) if we go this route.

maxme replied:
Done in 51dccbb - note the "Schedule with confidence" promo dialog was never showing before.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -855,73 +856,91 @@ public void onConfigurationChanged(Configuration newConfig) {
+            return PrimaryAction.SUBMIT_FOR_REVIEW;
+        }
+
+        if (mPost.isLocalDraft()) {
```
I just realized that a local draft can have a different post status:
- create a new post
- update the post status to "Private"
- we shouldn't offer the "Publish Now" option as the secondary action.

Since posts are created with the `Draft` status by default, I think it's fine to remove this test. I'm just wondering if we should differentiate "first" actions from other actions ("Update" seems weird for a Private post you just created)

maxme replied:
Removed in e1feffb

maxme commented:
Got @osullivanchris green light on paCBwp-95-p2 - this PR is now ready for review.

maxme commented:
> When I don't have permissions to publish posts, I can still see the "Submit" action on a Trashed post which moves the post back to drafts.

osullivanchris commented:
Sounds right, just to clarify related actions.

- For a trashed post **with** permissions, the action I'm seeing is **_'update'_**. This is the same action for a regular post **with** permissions
- Therefore, the actions on trashed posts **without** permissions here could be similarly consistent whether the post is trashed or not. Is the action for a **non-trashed** post **without** permissions **_'Save'_** as proposed here? If so, agree and go ahead

maxme commented:
> We might want to consider changing the target branch to a feature branch. Since we don't support remote auto-save yet and this solutions kind of counts on it. 

The updated primary/secondary actions work well even without auto-save. Not sure what would prevent us to merge this in `develop`.

> I'd also consider updating the releases notes as this change has direct impact on the users.

I'll update it.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -855,73 +857,91 @@ public void onConfigurationChanged(Configuration newConfig) {
+            case UNKNOWN:
+                return PrimaryAction.UPDATE;
         }
+        return PrimaryAction.SAVE;
```
Good point - I wanted to review all of these switch/cases and make them consistent before opening making this PR ready for review, but I forgot about it. Updated in 4d88bc2

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -855,73 +857,91 @@ public void onConfigurationChanged(Configuration newConfig) {
+            case SAVE:
+                return getString(R.string.save);
         }
+        return getString(R.string.save);
```
Updated in 4d88bc2

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -855,73 +857,91 @@ public void onConfigurationChanged(Configuration newConfig) {
+            case UNKNOWN:
+                return SecondaryAction.SAVE_AS_DRAFT;
         }
+        return SecondaryAction.NONE;
```
Updated in 4d88bc2

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -855,73 +857,91 @@ public void onConfigurationChanged(Configuration newConfig) {
+            case NONE:
+                throw new IllegalStateException("getSecondaryActionText shouldn't be called with that case");
+        }
+        throw new IllegalStateException("getSecondaryActionText shouldn't be called with that case");
```
Updated in 4d88bc2

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1435,6 +1409,55 @@ public void onClick(View view) {
+                showPublishConfirmationDialogAndPublishPost();
+                return true;
+            case NONE:
+                // Error: noop
```
Updated in 4d88bc2

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1484,29 +1507,55 @@ private void trackPostSessionEditorModeSwitch() {
+            // In other cases, we'll upload the post without changing its status
+            case SUBMIT_FOR_REVIEW:
+            case SAVE:
+                uploadPost(false);
```
Updated in 4d88bc2

maxme commented:
> For a trashed post with permissions, the action I'm seeing is 'update'. This is the same action for a regular post with permissions

- You see "Update" prior to the patch. 
- After the patch you'll see "Save" for trashed post with publishing permission.

> Therefore, the actions on trashed posts without permissions here could be similarly consistent whether the post is trashed or not. Is the action for a non-trashed post without permissions 'Save' as proposed here? If so, agree and go ahead

Non-trashed post primary action without permission (like draft or published post) isn't "Save", it's "Submit" (as "Submit for review").

maxme commented:
@osullivanchris Here is a [demo video](https://bia.is/s/AIDO/no-publishing-permissions-draft:submit:trash:recover.webm) of what I propose for users who don't have permission to publish.

osullivanchris commented:
Ah yes I follow now. Didn't realise this was in the same thread as our other actions updates. 

Can you confirm this is the behaviour after this change (want to make sure I understand):

**Drafts action**
With permission: Publish
Without permission: Submit for review

**Trashed action:**
With permission: Save (and returns it to drafts)
Without permission: Save (and returns it to drafts)

That makes sense to me! Everything behaves the same for a user without permissions except when they go to put something live, which is the appropriate difference.

maxme commented:
> Can you confirm this is the behaviour after this change

maxme commented:
@malinajirka this is ready for review

Updated release notes here: b46f9367a52df4e1c33e1e13b0d04fd73f12d651
Updated logic for user who don't have publishing permission here: 4591b45

malinajirka commented:
LGTM! Thanks Maxime! ;)

>We might want to consider changing the target branch to a feature branch. Since we don't support remote auto-save yet and this solutions kind of counts on it.
>The updated primary/secondary actions work well even without auto-save. Not sure what would prevent us to merge this in develop.

You are right. I thought we removed the "save" action for some of the flows, but it's not the case. It's everywhere where it needs to be.

🚢
